paasta_tools/apply_puppet_resources.py

```diff
@@ -0,0 +1,70 @@
+#!/opt/venvs/paasta-tools/bin/python
+import os
+import sys
+from filecmp import cmp
+from shutil import copy
+from subprocess import CalledProcessError
+from subprocess import run
+from traceback import print_exc
+
+APPLIED_DIRECTORY = ".applied"
+
+
+# This script expects the KUBECONFIG environment variable to be set correctly
+def main() -> int:
+    exit_code = 0
+    puppet_resource_root = sys.argv[1]
+    applied_resource_root = os.path.join(puppet_resource_root, APPLIED_DIRECTORY)
+
+    # Loop through everything in the puppet resource path
+    for root, dirs, files in os.walk(puppet_resource_root):
+        if APPLIED_DIRECTORY in dirs:
+            dirs.remove(APPLIED_DIRECTORY)
+
+        # Check to see if there's a difference between what Puppet created and
+        # what's been previously applied
+        for filename in [f for f in files if f.endswith(".yaml")]:
+            path = os.path.join(root, filename)
+            applied_path = os.path.join(
+                applied_resource_root, path[len(puppet_resource_root) + 1 :]
+            )
+            print(f"comparing {path} and {applied_path}")
+            if not os.path.exists(applied_path) or not cmp(
+                path, applied_path, shallow=False
+            ):
+                # This is idempotent; if something gets out of sync and a resource gets applied
+                # a second time, kubectl just won't make any changes
+                try:
+                    run(["kubectl", "apply", "-f", path], check=True)
+                    os.makedirs(applied_path[: -len(filename)], exist_ok=True)
+                    copy(path, applied_path)
```

Comment:
If the new version removes a field from what was previously there, I don't think this will catch it.

[this page](https://kubernetes.io/docs/tasks/manage-kubernetes-objects/declarative-config/#how-to-update-objects) implies that maybe `kubectl diff` will make that possible...?

Reply:
As best as I can tell, `kubectl apply` handles fields that get removed.